libs/native/jna/src/main/java/org/elasticsearch/nativeaccess/jna/JnaZstdLibrary.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.nativeaccess.jna;
+
+import com.sun.jna.Library;
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+
+import com.sun.jna.Pointer;
+
+import org.elasticsearch.nativeaccess.lib.ZstdLibrary;
+
+import java.nio.ByteBuffer;
+
+public class JnaZstdLibrary implements ZstdLibrary {
+
+    private interface NativeFunctions extends Library {
+        long ZSTD_compressBound(int scrLen);
+
+        long ZSTD_compress(Pointer dst, int dstLen, Pointer src, int srcLen, int compressionLevel);
+
+        boolean ZSTD_isError(long code);
+
+        String ZSTD_getErrorName(long code);
+
+        long ZSTD_decompress(Pointer dst, int dstLen, Pointer src, int srcLen);
+    }
+
+    private final NativeFunctions functions;
+
+    JnaZstdLibrary() {
+        this.functions = Native.load("zstd", NativeFunctions.class);
+    }
+
+    @Override
+    public long compressBound(int scrLen) {
+        return functions.ZSTD_compressBound(scrLen);
+    }
+
+    @Override
+    public long compress(ByteBuffer dst, int dstLen, ByteBuffer src, int srcLen, int compressionLevel) {
+        assert dst.isDirect() == false;
+        assert src.isDirect() == false;
+        assert dstLen != 0;
+        assert srcLen != 0;
+        try (Memory nativeDst = new Memory(dstLen);
+             Memory nativeSrc = new Memory(srcLen)) {
+            nativeSrc.write(0, src.array(), src.position(), srcLen);
```

Comment:
You already have asserts that the buffer is not direct, but since you're using `array()` here, I guess that the buffer needs to not be read-only either. I wonder how much work that would be to make it work transparently across direct, heap and read-only byte buffers. (could be a follow-up)

Reply:
I think I can make it work, it just needs the correct ByteBuffer calls to be backing agnostic.